src/Penrose/Substance.hs

```diff
@@ -83,6 +83,10 @@ instance Show Predicate where
         where nString = show predicateName
               aString = show predicateArgs
 
+-- | Special construct added for "declare and bind" statements, which get parsed into a declare and then a bind in the Substance core language
```

Comment:
I would generally avoid nesting the AST like this for (1) readability and (2) easier post-processing. In addition, I thought only `Expr`s are allowed on the right-hand side, so It would be cleaner to add `DeclBind T Var Expr` to the `SubStmt` type.

Reply:
In principle I would actually prefer both `DeclBind` and `DeclList` to be separate syntactic forms from `SubStmt` because they are convenience statements that aren't part of "Substance Core" (i.e. can already be expressed by combinations of Substance Core statements). Pragmatically, if you still think `DeclBind` should go in `SubStmt`, then we can move it there because it makes the code simpler. What do you think?

Reply:
Unless we plan to do anything other than deconstruction to the non-core statements, I think it would be better to keep it simple and avoid the unnecessary code. In general, it's a bit easier to debug the parser if the AST structure is simpler. If such changes add any useful information, I would leave it in (e.g. making pretty printing easier). In this case, I don't think we are gaining anything more than an extra layer to unwrap later.

Reply:
Another potential issue with putting those extra forms into Substance Core is that if we are analyzing the Substance Core AST but always compiling the "extra" Substance forms into other Substance forms, then we will have two pattern-match branches that are always unused in the SubStmt type, which can lead to runtime errors.

Reply:
Again, I think the `DeclList` and `DeclBind` are convenience forms that are mainly for the user and parser, and should be kept separate from the core grammar of the language. I see that `refineDeclList` is basically doing the same thing with turning the `DeclList` convenience form into a list of declarations and so I would rather move `DeclList` to `SubStmt'` as well.

So, I think I am not personally going to revert these Substance changes. However, if it's important to you as someone who spends more time working with the parser (e.g. if the different Substance form affects the IDE typeahead), then you should go ahead and revert the changes and I won't complain further about it 🙂